### PR TITLE
Revert and deprecate StakeHelper

### LINKED ISSUE #81: Revert stake helper changes
The stake helper constructor should be changed back to the old interface and the whole class should be marked as deprecated.

### PR DESCRIPTION
**Pull request by schemar**
I reverted the changes to the constructor of the StakeHelper.
I also marked the StakeHelper as deprecated and added deprecation warnings when StakeHelper is used.

Fixes #81

### CODE REVIEW

**Review by deepesh-kn — CHANGES_REQUESTED** [review]
I thought I will be doing this ticket.

Here is the reference of the original file.
https://github.com/OpenSTFoundation/mosaic-tbd/blob/master/libs/helpers/StakeHelper.js

functions not in use form these files are:
- perform
- getBounty
- approveStakeAmount
- stake

functions that are being used:
- constructor // with not validation, no params are passed.
- getNonce
- _approveStakeAmountRawTx
- _stakeRawTx
- createSecretHashLock
- toHashLock

So on the current file for this repository, we need the following changes.
- Change the constructor like the original file (no validation on constructor params).
- Rename `_getNonce` to `getNonce` and delete exisiting getNonce
- Delete getBounty
- Delete getValueToken
- Delete approveStakeAmount

**Comment by schemar** [comment]
> I thought I will be doing this ticket.
> 
> Here is the reference of the original file.
> https://github.com/OpenSTFoundation/mosaic-tbd/blob/master/libs/helpers/StakeHelper.js
> 
> functions not in use form these files are:
> 
>     * perform
> 
>     * getBounty
> 
>     * approveStakeAmount
> 
>     * stake
> 
> 
> functions that are being used:
> 
>     * constructor // with not validation, no params are passed.
> 
>     * getNonce
> 
>     * _approveStakeAmountRawTx
> 
>     * _stakeRawTx
> 
>     * createSecretHashLock
> 
>     * toHashLock
> 
> 
> So on the current file for this repository, we need the following changes.
> 
>     * Change the constructor like the original file (no validation on constructor params).
> 
>     * Rename `_getNonce` to `getNonce` and delete exisiting getNonce
> 
>     * Delete getBounty
> 
>     * Delete getValueToken
> 
>     * Delete approveStakeAmount

OK, thank you for the help. What about `perform` and `stake`?

**Comment by schemar** [comment]
@deepesh-kn I am in the process of reverting it to https://raw.githubusercontent.com/OpenSTFoundation/mosaic-tbd/master/libs/helpers/StakeHelper.js

The constructor is completely different.

**Comment by schemar** [comment]
I will delete the tests that we wrote as they don't make sense anymore.

**Comment by schemar** [comment]
Ready for review again (?)

**Review by deepesh-kn — APPROVED** [review]
LGTM 🎢 
Reverting to original looks good to me.

**Comment by schemar** [comment]
Ready for review again @deepesh-kn @hobofan . I added the integration tests for the stake helper and the eip20 token in the integration tests.

**Comment by schemar** [comment]
Reminder: add „use strict“

File: src/helpers/StakeHelper.js [hunk 1]
```diff
@@ -1,267 +1,300 @@
...
-   * @param {string} nonce Staker nonce.
-   * @param {string} hashLock Hash lock.
-   * @param {Object} txOptions Transaction options.
-   * @param {Object} web3 Web3 object.
-   * @param {string} gateway EIP20Gateway contract address.
-   * @param {string} staker Staker address.
-   *
-   * @returns {Object} Raw transaction object.
-   */
+  stake(
```
Comment (deepesh-kn) [inline src/helpers/StakeHelper.js:196]: May be we can retain the documentation

File: test_integration/03_stake/01_stake_helper.js [hunk 2]
```diff
@@ -0,0 +1,101 @@
+'use strict';
+
+const { assert } = require('chai');
+const BN = require('bn.js');
+const StakeHelper = require('../../src/helpers/StakeHelper');
+const shared = require('../shared');
+
+const validateReceipt = (receipt) => {
```
Comment (deepesh-kn) [inline test_integration/03_stake/01_stake_helper.js]: We are having the common functions in all the integration test files.
Maybe we can move it to a common file. (We can create a new ticket for this)

**Review by deepesh-kn — APPROVED** [review]
LGTM 🎢 
I have added 2 comments, you can check if it should be done or not.

File: src/helpers/StakeHelper.js (hunk 1 above)
Comment (schemar) [reply src/helpers/StakeHelper.js:196]: I would need to write  new documentation, right?

File: test_integration/03_stake/01_stake_helper.js (hunk 2 above)
Comment (schemar) [reply test_integration/03_stake/01_stake_helper.js]: I will create a ticket. This could potentially go into a separate utils repo or something like that.

Comment (schemar) [reply test_integration/03_stake/01_stake_helper.js]: https://github.com/OpenSTFoundation/mosaic.js/issues/98

File: src/helpers/StakeHelper.js (hunk 1 above)
Comment (deepesh-kn) [reply src/helpers/StakeHelper.js:196]: Yeah !!! Maybe we can leave it for now as its deprecated.

File: index.js [hunk 3]
```diff
@@ -1,31 +1,96 @@
 'use strict';
 
-const Mosaic = require('./src/Mosaic');
-const Chain = require('./src/Chain');
+const MosaicProvider = require('./src/Mosaic');
+
 const AbiBinProvider = require('./src/AbiBinProvider');
+const Chain = require('./src/Chain');
 const ChainSetup = require('./src/ChainSetup');
-const Contracts = require('./src/Contracts');
```
Comment (0xsarvesh) [inline index.js]: I think, BrandedToken.js still consumes Contract class. This should be exposed.

File: index.js [hunk 4]
```diff
@@ -1,31 +1,96 @@
...
+
+class Helpers {
+  static get StakeHelper() {
+    Utils.deprecationNoticeStakeHelper();
+
+    return StakeHelper;
+  }
+}
+
+class Mosaic {
```
Comment (0xsarvesh) [inline index.js]: From consumer point of view, `Mosaic.Mosaic` seems little confusing 🤔 
```
Const Mosaic = require('@openstfoundation/mosaic.js');

const mosaicProvider = new Mosaic.Mosaic(origin, auxiliary);
```

File: src/helpers/StakeHelper.js [hunk 5]
```diff
@@ -1,267 +1,300 @@
...
+      _gasPrice,
+      _gasLimit,
+      _nonce,
+      _hashLock,
+      txOptions,
+      web3,
+      gateway,
+      staker,
+    );
+    console.log(`* Staking SimpleToken`);
```
Comment (0xsarvesh) [inline src/helpers/StakeHelper.js:228]: Though, it's deprecated. We should remove `console.log` and keep documentation.

File: test_integration/03_stake/01_stake_helper.js [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+    // Only for testing purposes we re-use the same address:
+    const beneficiary = shared.setupConfig.deployerAddress;
+    const gasPrice = '0';
+    const gasLimit = '10000000';
+
+    return helper
+      .perform(amountToStake, beneficiary, gasPrice, gasLimit)
+      .then((output) => {
+        validateReceipt(output.receipt);
```
Comment (0xsarvesh) [inline test_integration/03_stake/01_stake_helper.js]: Asserting that receipt is valid is sufficient? 
Or it should also assert other parameters like `hash lock`, `amount` etc in the event?

**Review by 0xsarvesh — APPROVED** [review]
LGTM 👍 💯 
Some feedbacks inline.

File: test_integration/03_stake/01_stake_helper.js (hunk 6 above)
Comment (schemar) [reply test_integration/03_stake/01_stake_helper.js]: I only ported the tests from `tests`. I can add checks. Should I?

File: src/helpers/StakeHelper.js (hunk 5 above)
Comment (schemar) [reply src/helpers/StakeHelper.js:228]: We cannot "keep" it. It doesn't exist. We would need to write it.

File: index.js (hunk 4 above)
Comment (schemar) [reply index.js]: Agreed. But that was a different PR: #44 
Should we rename it now before the beta release?

@hobofan @deepesh-kn

File: index.js (hunk 3 above)
Comment (schemar) [reply index.js]: That sucks. Now we have two kinds of "contracts". `Contracts` and `ContractInteracts`...
Can we call `ContractInteract` ↦ `Contract` and `Contract` ↦ `SomethingElse`??

@deepesh-kn @hobofan

File: index.js (hunk 4 above)
Comment (hobofan) [reply index.js]: @schemar I think you also made the same remark when writing this and we had a discussion where we ended up with the solution that we **keep it like this for now** and then make `Mosaic` the top level class object that's exposed from the library, and it has static accessor functions for the other parts of the library.

File: index.js (hunk 3 above)
Comment (schemar) [reply index.js]: expose contract and contract interact separately. Create follow-up ticket to reevaluate `contracts`

File: index.js (hunk 4 above)
Comment (schemar) [reply index.js]: Pull up the mosaic class. Classes uppercase, objects lowercase.

File: src/helpers/StakeHelper.js (hunk 5 above)
Comment (schemar) [reply src/helpers/StakeHelper.js:228]: keep file as we received it

File: test_integration/03_stake/01_stake_helper.js (hunk 6 above)
Comment (schemar) [reply test_integration/03_stake/01_stake_helper.js]: no tests for deprecated stake helper

File: index.js (hunk 3 above)
Comment (schemar) [reply index.js]: https://github.com/OpenSTFoundation/mosaic.js/issues/111

**Review by hobofan — APPROVED** [review]
LGTM!

File: test_integration/contracts/EIP20Token.json [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
...94909404938401819004810282018101909252828152606093909290918301828280156103515780601f1061032657610100808354040283529160200191610351565b60006104c333848461056c565b9392505050565b600160a060020a03918216600090815260056020908152604080832093909416825291909152205490565b60008282111561056657604080517f08c379a000000000000000000000000000000000000000000000000000000000815260206004820152601b60248201527f556e646572666c6f77207768656e207375627472616374696e672e0000000000604482015290519081900360640190fd5b50900390565b600160a060020a038316600090815260046020526040812054610595908363ffffffff6104f516565b600160a060020a0380861660009081526004602052604080822093909355908516815220546105ca908363ffffffff61062b16565b600160a060020a0380851660008181526004602090815260409182902094909455805186815290519193928816927fddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef92918290030190a35060019392505050565b6000828201838110156104c357604080517f08c379a000000000000000000000000000000000000000000000000000000000815260206004820152601560248201527f4f766572666c6f77207768656e20616464696e672e0000000000000000000000604482015290519081900360640190fdfea165627a7a723058205d93888008fa3996759fca88e66232a119e141917a86569af7e5cd7b6a096cb40029",
```
Comment (deepesh-kn) [inline test_integration/contracts/EIP20Token.json:2]: Why is this filed added? We can get the abi and bin from the npm?

Comment (hobofan) [reply test_integration/contracts/EIP20Token.json:2]: I am assuming a that this is a "real" EIP20Token, in contrast to the one we are currently exposing from mosaic-contracts. See also https://github.com/OpenSTFoundation/mosaic-contracts/issues/629

Comment (schemar) [reply test_integration/contracts/EIP20Token.json:2]: It is. It is not part of `mosaic-contracts`.

**Comment by hobofan** [comment]
👍

**Review by 0xsarvesh — APPROVED** [review]
LGTM after changes. 🙌 🚀 🏁

**Pull request merged**